internal/rbd/replicationcontrollerserver.go

```diff
@@ -265,12 +292,27 @@ func (cs *ControllerServer) PromoteVolume(ctx context.Context,
 // volumeID, If the image is present, mirroring is enabled and the
 // image is in promoted state it will demote the volume as secondary.
 // If the image is already secondary it will return success.
-func (cs *ControllerServer) DemoteVolume(ctx context.Context,
+func (rs *ReplicationServer) DemoteVolume(ctx context.Context,
 	req *replication.DemoteVolumeRequest,
 ) (*replication.DemoteVolumeResponse, error) {
-	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
-	defer cs.cleanup(rbdVol, cr)
+	volumeID := req.GetVolumeId()
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "empty volume ID in request")
+	}
+	rbdVol, cr, err := rs.getVolumeFromID(ctx, volumeID, req.GetSecrets())
+	defer rs.cleanup(rbdVol, cr)
 	if err != nil {
+		switch {
+		case errors.Is(err, util.ErrOperationExists):
+			util.ErrorLog(ctx, util.VolumeOperationAlreadyExistsFmt, volumeID)
+			err = status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeID)
+		case errors.Is(err, ErrImageNotFound):
+			err = status.Errorf(codes.NotFound, "volume %s not found", volumeID)
```

Comment:
just thinking, `volume not found` or `image not found` is better here?

Reply:
as we are considering image as the volume, IMO volume should be fine here

Reply:
all the mirror operations log ( rbd_mirror*,)pointing/referencing `image` and here its volume or mix (like in DisableVolumeReplication()) at times. in case of rbd, volume and image is pretty much same, so I was thinking the uniqueness is good

Reply:
as it's a volumeID which we got from the request, sending an error as the volume Not found. all the functions in mirror.go are referring to the image because after decoding the volumeID we got the image name as those are internal functions and we are logging it as an image. Do you still think the error message we send back to the RPC calls should say `image not found` instead of `volume not found` , if yes,i can do it.

Reply:
bit difficult while we interchange the reference, but its fine for now .